### ISSUE TITLE
Feature: PostgreSQL as a Queue

### ISSUE THREAD

**Issue by Vad1mo** [airtai/faststream#799]
**Is your feature request related to a problem? Please describe.**

I am aware that PostgreSQL is now a Message Queue per se; however, it can be utilized as such. Many developers realize that it suffices for the majority of tasks, and it comes with zero added costs if a PostgreSQL db is present.

**Describe the solution you'd like**
Clearly and concisely describe the desired outcome or solution.

**Feature code example**
Nothing in particular 

**Describe alternatives you've considered**
You can do exactly this with MySQL and SQL server too because they both support SKIP LOCKED.

**Additional context**

- https://adriano.fyi/posts/2023-09-24-choose-postgres-queue-technology/
- https://www.prequel.co/blog/sql-maxis-why-we-ditched-rabbitmq-and-replaced-it-with-a-postgres-queue
- https://dagster.io/blog/skip-kafka-use-postgres-message-queue

**Comment by acaloiaro** [comment]
@Vad1mo :wave: I came across this by chance and see you've referenced my write up on the topic of PG as a queue. 

If you find Go easy enough to read, you might benefit from neoq (https://github.com/acaloiaro/neoq) as a reference implementation for the ideas lain out in my post.

Cheers!

**Comment by Vad1mo** [comment]
> @Vad1mo :wave: I came across this by chance and see you've referenced my write up on the topic of PG as a queue. 
> 
> 
> 
> If you find Go easy enough to read, you might benefit from reading the neoq source (https://github.com/acaloiaro/neoq) as a reference implementation for the ideas lain out in my post. The whole PG implementation is contained within: https://github.com/acaloiaro/neoq/blob/main/backends/postgres/postgres_backend.go
> 
> 
> 
> Cheers!

Found it already 👀 https://github.com/acaloiaro/neoq/issues/70

**Comment by Lancetnik** [comment]
I don't like an idea to poll DB for updates
I am thinking about Listen/Notify PG feature or even CDC (replication journal) consuming

**Comment by Vad1mo** [comment]
Yes, I was referring to listen/notify. No mention about polling. 
Regarding CDC, here I am not sure how or if it is low level, hence if it can be easily setup within the offerings of cloud providers.

**Comment by Lancetnik** [comment]
CDC available from any pg driver: asyncpg/psycopg/etc
It is just a part of protocol
So, it can be a useful feature as I can image

**Comment by mixartemev** [comment]
Hey guys! I really need it asap. R u planning to implement it in nearest future or should I search other tools for manage pg listen/notify queues?

**Comment by Lancetnik** [comment]
> Hey guys! I really need it asap. R u planning to implement it in nearest future or should I search other tools for manage pg listen/notify queues?

Sorry, I don't think, that we can add Postgres support this year. Probably, we can do it in the first half of 2025, but can't strongly promise 😢

**Comment by mcproger** [comment]
Hey @Lancetnik

I've built a simple PoC with the listen/notify approach (took a couple of hours, so not the best code quality) that shows the general idea – [link](https://github.com/mcproger/pg-pub-sub/blob/becff54b2381b9bfeed926ff5712923f08031bbc/src/listener.py#L48)

If you are OK with this way, I would like to try to implement this feature. I'm still tinkering with the source code so if you can point places to start with it would be very helpful (if no - not a problem, just will spend a little bit more time reading the code).

Thanks

**Comment by bakhoraliev** [comment]
Hi @Lancetnik, if I add a fully compatible implementation of PostgreSQL as a broker, will it be merged, or you not see it in the core?

**Comment by Lancetnik** [comment]
@bakhoraliev great news! I think, you are talking about – https://github.com/bakhoraliev/faststream-postgresql
Let me check the repo before make a decision.

**Comment by bakhoraliev** [comment]
@Lancetnik, I recently created fork and started working on it, but I caught myself thinking that if you don't see PostgreSQL as broker in the core, is it worth spending time on it? Currently, there are no additional commits in the fork. But if you say that it could be in the core, the work will be done soon.

**Comment by Lancetnik** [comment]
@bakhoraliev I think, we can merge it to the project. Are you using Listen / Notify mechanism or smth else?

**Comment by bakhoraliev** [comment]
@Lancetnik, we have several ways to implement PostgreSQL as a broker: using the Listen/Notify mechanism, reading WAL, or polling the database, but each method has its pros and cons.

**Listen/Notify**
Pros: 
- no additional infrastructure needed
Cons: 
- if client is not listening when notification is sent, the message is lost
- limited payload size (it must be shorter than 8000 bytes)
- no horizontal scaling, because if two instances Listen same channel, they receive same notifications

**reading WAL**
Pros:
- captures all database changes reliably
- easy to set up and use
Cons:
- requires additional tooling(wal2json) to transform WAL changes into messages
- no horizontal scaling, because if two instances read WAL changes, they receive same logs

**database polling**
Pros:
- simple to implement using usual SQL queries
- can guarantee reliable message processing based on transactional data
- messages persistence
Cons:
- adds latency and unnecessary load due to periodic querying

My proposal: start with database polling and implement this solution first, as it covers all cases.
Next step will be to combine Listen/Notify mechanism with database polling. This synthesis will work as follows:
1. First start of the consumer SELECT's unprocessed messages from the database and process them
2. All new messages will be processed by notifications from Notify, but the payload will contain pk from table. The consumer will SELECT messages and add advisory block. This will give us the ability to scale horizontally.
3. Processed messages will be deleted (maybe soft-deleted)

**Comment by Lancetnik** [comment]
@bakhoraliev I think, we can support all cases in the way we did with NATS, Redis and other broker. Smth like:
* `@broker.subscriber(table="...")`
* `@broker.subscriber(listen="...")`
* `@broker.subscriber(wal="...")`

But, start with polling sounds reasonable

**Comment by bakhoraliev** [comment]
@Lancetnik, thank you for your reply. I will start with database polling. See you soon.

**Comment by DABND19** [comment]
@bakhoraliev Hi! Any updates? Can we collaborate together?

**Comment by bakhoraliev** [comment]
@DABND19, Hi, to be honest, I'm working on an In-Process broker right now because while I was trying to figure out how broker integrations were done, I realized it would be easier to do an In-Process broker first.

**Comment by alicja-karasiewicz** [comment]
Hey, any updates? :) Do you maybe need some help with this feature?

**Comment by heldergomes** [comment]
Hi, any updates?
I'm interested in this CR!

**Comment by Lancetnik** [comment]
Hi! To be honest, PostgreSQL support isn't on our roadmap.

My current thinking for 1.0.0 is to keep a limited set of brokers supported in the core - just the most common ones: RMQ, Kafka, NATS, Redis, MQTT, SQS, and ZMQ. Support for other brokers (including PostgreSQL) will be community-driven. The project is maintained by a fairly small team, so we simply don't have the resources to support every broker ourselves.

That's exactly why stable broker development APIs are one of the remaining items before 1.0 - once they land, building and maintaining a third-party broker integration should be much more approachable, and we'd be happy to point to community-maintained ones from the docs.

We'll reopen the issue if it gains strong community support!

**Comment by Vad1mo** [comment]
If there are interfaces for that, happy to contribute a community PG implementation

**Comment by Lancetnik** [comment]
Great news! We're already planning a release of such a broker — it will live as a community-maintained package at https://github.com/faststream-community/faststream-sqlbroker.

@Arseniy-Popov (the author of #2704) has done great foundational work on this. If you'd like to contribute, feel free to reach out to them in the context of the new repository!
